pandas/_libs/lib.pyx

```diff
@@ -2357,3 +2354,115 @@ def fast_multiget(dict mapping, ndarray keys, default=np.nan):
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef void _concat_date_cols_numpy(tuple date_cols, object[:] result_view,
+                                  Py_ssize_t rows_count, Py_ssize_t col_count,
+                                  bint keep_trivial_numbers):
+    cdef:
+        Py_ssize_t col_idx, row_idx
+        list list_to_join
+        cnp.ndarray[object] iters
+        object[::1] iters_view
+        flatiter it
+
+    if col_count == 1:
+        array = date_cols[0]
+        it = <flatiter>PyArray_IterNew(array)
+        for row_idx in range(rows_count):
+            item = PyArray_GETITEM(array, PyArray_ITER_DATA(it))
+            convert_and_set_item(item, row_idx, result_view,
+                                 keep_trivial_numbers)
+            PyArray_ITER_NEXT(it)
+    else:
+        list_to_join = [None] * col_count
+        # setup iterators
+        iters = np.zeros(col_count, dtype=object)
+        iters_view = iters
+        for col_idx, array in enumerate(date_cols):
+            iters_view[col_idx] = PyArray_IterNew(array)
+        for row_idx in range(rows_count):
+            for col_idx, array in enumerate(date_cols):
+                it = <flatiter>iters_view[col_idx]
+                item = PyArray_GETITEM(array, PyArray_ITER_DATA(it))
+                put_object_as_unicode(list_to_join, col_idx, item)
+                PyArray_ITER_NEXT(it)
+            result_view[row_idx] = PyUnicode_Join(' ', list_to_join)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef void _concat_date_cols_sequence(tuple date_cols, object[:] result_view,
+                                     Py_ssize_t rows_count,
+                                     Py_ssize_t col_count,
+                                     bint keep_trivial_numbers):
+    cdef:
+        Py_ssize_t col_idx, row_idx
+        list list_to_join
+
+    if col_count == 1:
+        for row_idx, item in enumerate(date_cols[0]):
+            convert_and_set_item(item, row_idx, result_view,
+                                 keep_trivial_numbers)
+    else:
+        list_to_join = [None] * col_count
+        for row_idx in range(rows_count):
+            for col_idx, array in enumerate(date_cols):
+                put_object_as_unicode(list_to_join, col_idx, array[row_idx])
+            result_view[row_idx] = PyUnicode_Join(' ', list_to_join)
+
+
+def _concat_date_cols(tuple date_cols, bint keep_trivial_numbers=False):
+    cdef:
+        Py_ssize_t rows_count = 0, col_count = len(date_cols)
+        cnp.ndarray[object] result
+
+    if col_count == 0:
+        return np.zeros(0, dtype=object)
+
+    rows_count = min(len(array) for array in date_cols)
+
+    result = np.zeros(rows_count, dtype=object)
+    if all(util.is_array(array) for array in date_cols):
+        # call specialized function to increase performance
+        _concat_date_cols_numpy(date_cols, result, rows_count, col_count,
+                                keep_trivial_numbers)
```

Comment:
I would rather each of these routines actually allocate the memory rather than passing it in, and return the new object

Reply:
Done.